h2o-core/src/main/java/water/webserver/H2OHttpViewImpl.java

```diff
@@ -0,0 +1,194 @@
+    SERVLETS.put("/", RequestServer.class);
+  }
+
+  public H2OHttpViewImpl(H2OHttpConfig config) {
+    this.config = config;
+  }
+
+  public void acceptRequests() {
+    _acceptRequests = true;
+  }
+
+  /**
+   * @return URI scheme
+   */
+  public String getScheme() {
+    return config.jks == null ? "http" : "https";
+  }
+
+  @Override
+  public LinkedHashMap<String, Class<? extends HttpServlet>> getServlets() {
+    return SERVLETS;
+  }
+
+  @Override
+  public boolean authenticationHandler(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (!config.loginType.isJaas()) {
+      //TODO question: for LoginType.HASH, does this equal not adding this handler at all? if so, it might be better doing it that way
+      return false;
+    }
+
+    final String loginName = request.getUserPrincipal().getName();
+    if (loginName.equals(config.user_name)) {
+      return false;
+    }
+    Log.warn("Login name (" + loginName + ") does not match cluster owner name (" + config.user_name + ")");
+    ServletUtils.sendResponseError(response, HttpServletResponse.SC_UNAUTHORIZED, "Login name does not match cluster owner name");
+    return true;
+  }
+
+  @Override
+  public void gateHandler(HttpServletRequest request, HttpServletResponse response) {
+    ServletUtils.startRequestLifecycle();
+    while (!_acceptRequests) {
+      try { Thread.sleep(100); }
+      catch (Exception ignore) {}
+    }
+
+    boolean isXhrRequest = false;
+    if (request != null) {
+      isXhrRequest = ServletUtils.isXhrRequest(request);
+    }
+    ServletUtils.setCommonResponseHttpHeaders(response, isXhrRequest);
+  }
+
+  @Override
+  public H2OHttpConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public Collection<RequestAuthExtension> getAuthExtensions() {
+    return ExtensionManager.getInstance().getAuthExtensions();
+  }
+
+
+  // normal login handler part //todo: consider using mostly the same code as in proxy part below
+
+  @Override
+  public boolean loginHandler(String target, HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (! isLoginTarget(target)) {
+      return false;
+    }
+
+    if (isPageRequest(request)) {
+      sendLoginForm(request, response);
+    } else {
+      ServletUtils.sendResponseError(response, HttpServletResponse.SC_UNAUTHORIZED, "Access denied. Please login.");
+    }
+    return true;
+  }
+
+  private static void sendLoginForm(HttpServletRequest request, HttpServletResponse response) {
+    final String uri = ServletUtils.getDecodedUri(request);
+    try {
+      byte[] bytes;
+      try (InputStream resource = water.init.JarHash.getResource2("/login.html")) {
+        if (resource == null) {
+          throw new IllegalStateException("Login form not found");
+        }
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        water.util.FileUtils.copyStream(resource, baos, 2048);
```

Comment:
Why 2k ? Just asking.

Reply:
no idea - this code was copied as is, without modification, because I don't want to debug it now. If I do any change here, it will be replacing this code with one of many implementations of IOUtil.copy() already bundled in H2O. But preferably, not in this PR. Same applies to closing `baos`.

Reply:
Sure. No problem. Just pointed it out as I ran across it.